File: part1_syntax_semantics/section2/type_and_scope.py
# Python: dynamic typing + closures + scope demo
def make_counter(start=0):
    count = start          # 'count' lives in the enclosing (closure) scope
    def increment(step=1):
        nonlocal count     # explicit keyword required to write to the outer variable
        count += step
        return count
    return increment

counter = make_counter(10)
print("counter():", counter())   # 11
print("counter():", counter())   # 12
print("counter(5):", counter(5)) # 17

# Dynamic typing: the same variable can hold different types at different times
value = 5
print("value:", value, type(value))
value = "five"
print("value:", value, type(value))
value = [5]
print("value:", value, type(value))

# Duck typing: no interface is declared, Python only checks behavior at call time
def describe_length(x):
    return len(x)  # works for anything with __len__, fails at runtime otherwise

print("len of list:", describe_length([1, 2, 3]))
print("len of str:", describe_length("hello"))
try:
    describe_length(42)   # int has no __len__: raises at runtime, not compile time
except TypeError as e:
    print("TypeError caught:", e)

File: part1_syntax_semantics/section2/type_and_scope.js
// JavaScript: dynamic typing + closures + scope + var/let hoisting demo
function makeCounter(start = 0) {
    let count = start;          // captured by closure
    return function increment(step = 1) {
        count += step;          // no special keyword needed to mutate outer variable
        return count;
    };
}

const counter = makeCounter(10);
console.log("counter():", counter());   // 11
console.log("counter():", counter());   // 12
console.log("counter(5):", counter(5)); // 17

// Dynamic typing + implicit coercion (a JS-specific quirk)
let value = 5;
console.log("value:", value, typeof value);
value = "five";
console.log("value:", value, typeof value);
value = [5];
console.log("value:", value, typeof value);

console.log("1 + '1' =", 1 + "1");   // "11" -- number coerced to string
console.log("1 == '1' :", 1 == "1"); // true  -- loose equality coerces
console.log("1 === '1' :", 1 === "1"); // false -- strict equality does not

// var (function-scoped) vs let (block-scoped)
function scopeDemo() {
    if (true) {
        var fnScoped = "I leak out of the block";
        let blockScoped = "I stay inside the block";
    }
    console.log(fnScoped); // works: var ignores block scope
    try {
        console.log(blockScoped); // ReferenceError: block-scoped, not visible here
    } catch (e) {
        console.log("ReferenceError caught:", e.message);
    }
}
scopeDemo();
